### ISSUE TITLE
[CP/M] (classic): argc, argv[] setup exceeds the argument count

### ISSUE THREAD

**Issue by zx70** [z88dk/z88dk#1784]
When an argument is passed to the command line and the file redirection is used, the argument count overflows and picks a non-existing parameter.
```
#include <stdio.h>

int x;

int main(int argc, char *argv[])
{
	printf("Arg test\n");
	printf("\n");
	
		printf ("argc: %u\n",argc);
		
	for (x=0; x<argc; x++)
		printf ("argv[%u]: %s\n",x,argv[x]);
}
```

When run with no arguments it works as expected, but when if we run:
C:\> cpm test "arg1 arg2 arg3 arg4 >test"

we get:
```
Arg test

argc: 6
argv[0]:
argv[1]: 
argv[2]: ARG1
argv[3]: ARG2
argv[4]: ARG3
argv[5]: ARG4
```

The problem happens only when a single redirection option is used, when both the inbound and outbound redirection options are set, the issue does not show up

**Comment by zx70** [comment]
sorry, the issue was local, updating z88dk to the latest nightly build solved the problem

**Comment by zx70** [comment]
the culprit seems to be the fancy way I solved a previous bug  <blush> !
After the parameter loop I check if the first argument has been forgotten. .  better not to loose it at all..

**Comment by zx70** [comment]
It seems to work well, now.
@suborb,   I used LDDR.. is it ok to have it in the classic CP/M startup code or it will hit the 8080 targets?

**Comment by suborb** [comment]
Lddr will be emulated on 8080 targets (I hope!)

**Comment by zx70** [comment]
+new option to specify an optional program name for argv[0]
#pragma string name pgmname

**Comment by suborb** [comment]
This seems to have broken non-redirected argc, argv handling.

**Comment by zx70** [comment]
how frustrating  😖

**Comment by suborb** [comment]
Yeah. I think I’m going to revert to the original code since redirection isn’t often used.

**Comment by zx70** [comment]
mm ok, just do.   I'll have a crack on the current version and try to fix.

**Comment by zx70** [comment]
oddly I couldn't reproduce the problem on my current copy.   How did you spot it ?

**Comment by suborb** [comment]
Came out of this comment here: https://z88dk.org/forum/viewtopic.php?p=19715&sid=cb29a6710f1f13ba68956cdb30582c44#p19715

argv[1] ended up being garbage for CP/M. For ticks argv[1] had the first character cut off.

All a bit odd...

**Comment by zx70** [comment]
thanks for explaining,  I was checking argc only
